appendix-basic65.tex

```diff
@@ -220,168 +252,240 @@ \subsection{BASIC 65 Variables}
-and screen colour memory.
-
-\screentextwide{ T@\&(COLUMNS,ROWS)} is mapped to the text screen memory.
-
-\screentextwide{ C@\&(COLUMNS,ROWS)} is mapped to the colour screen memory.
-
-In 80 $\times$ 25 text mode the arrays are dimensioned as:
-
-\screentextwide{ T@\&(79,24)} Column = 0 - 79, Row = 0 - 24.
-
-\screentextwide{ C@\&(79,24)} Column = 0 - 79, Row = 0 - 24.
-
-In 80 $\times$ 50 text mode the arrays are dimensioned as:
-
-\screentextwide{ T@\&(79,49)} Column = 0 - 79, Row = 0 - 49.
-
-\screentextwide{ C@\&(79,49)} Column = 0 - 79, Row = 0 - 49.
-
-In 40 $\times$ 25 text mode the arrays are dimensioned as:
-
-\screentextwide{ T@\&(39,24)} Column = 0 - 39, Row = 0 - 24.
-
-\screentextwide{ C@\&(39,24)} Column = 0 - 39, Row = 0 - 24.
-
-If you want to display the character {\bf A} (display code = 1)
-in row 2 and column 5 and make it yellow, use the mapped byte arrays:
-
-\screentextwide{T@\&(5,2) = 1} character A
-
-\screentextwide{C@\&(5,2) = 7} colour YELLOW
-
-Examining the contents of a screen position is equally easy.
-To get the character and the colour at the left most (home) postion:
-
-\screentextwide{T\% = T@\&(0,0)}
-
-\screentextwide{C\% = C@\&(0,0)}
-
-T\% will then contain the display code of the character, while C\% has the colour index.
-
+A variable that stores a single value is also known as a {\em scalar
+variable}. Commodore BASIC also supports a type of variable that can store
+multiple values, called an {\em array}.
+
+The following example stores three string values in an array, then uses a {\bf
+FOR} loop to {\bf PRINT} a message for each element:
+
+\begin{screenoutput}
+10 DIM NA$(3)
+20 NA$(0) = "DEFT"
+30 NA$(1) = "GARDNERS"
+40 NA$(2) = "LYDON"
+50 FOR I=0 TO 3
+60 PRINT "HELLO, ";NA$(I);"!"
+70 NEXT I
+\end{screenoutput}
+
+Each value in an array is referenced by the name of the array variable and an
+integer index. For example, \screentext{AA(7)} refers to the element of the
+array named \screentext{AA()} with index 7. Indexes are ``zero-based:'' the
+first element in the array has an index of 0. The index can be a numeric
+expression, which can be a powerful way to operate on multiple elements of
+data.
+
+All values in an array must be of the same type. The type is indicated in the
+name of the variable, similar to scalar variables. \screentext{AA()} is an
+array of real numbers, \screentext{AA\$()} is an array of strings.
+
+Array variable names are considered separate from scalar variable names. The
+scalar variable \screentext{AA} has no relationship to the array variable
+\screentext{AA()}.
+
+Array variables must be declared before use with their maximum size. BASIC 65
+allocates the maximum amount of memory needed for the array. The {\bf DIM}
+command declares (``dimensions'') an array.
+
+An array can have multiple dimensions, each with its own index separated by a
+comma. The array must be declared with the maximum value for each dimension.
```

Comment:
Bit Shifter's old text mentioned how undeclared arrays would get automatically declared with an upper boundary of 10

It's an interesting tidbit, maybe worth keeping a mention of it?

Reply:
Rewrote this paragraph to make that explicit, thanks! It's more than just interesting because it's a potential source of confusion: if you use an array variable before DIM, the array is allocated with 10 elements, but then the DIM fails because it's a re-dimensioning of the array.